### DISCUSSION TITLE
Different dependencies for different CMake targets (e.g. test and production)

### DISCUSSION THREAD

**Discussion by lprc** [microsoft/vcpkg#33458] [Q&A] [answered]
I can't think of this question not being asked before, but I wasn't able to find any information on this, so here we go:

Is it not possible with vcpkg to have different dependencies for different CMake targets? 

There is an [example repo](https://github.com/miredirex/cmake-vcpkg-example) that uses vcpkg in manifest mode together with CMake, which uses catch2 and fmt as dependencies. However catch2 is obviously only needed for testing purposes. Even when removing the test-subdirectory from `lib`, i.e. removing the dependency of catch2 for the whole project, but leaving it as a dependency in `vcpkg.json`, it will still download catch2 in the configure step of cmake. This can lead to quite an overhead if there are several testing-only-dependencies but you only want to install the library of question, when sizes get bigger.

In my project I tried to put several `vcpkg.json` files in the subdirectories, each having a `CMakeLists.txt` file inside and getting added by `add_subdirectory` by the root CMakeLists.txt, but then apparently vcpkg does not download the dependencies.

Maybe I'm overseeing something here as I just started experimenting with vcpkg. Please tell me if this is the expected behaviour and/or if there is some workaround.

**Comment by quyykk** [comment, marked as answer]
Yes, you can use features:

```json
{
  "dependencies": [
    "fmt"
  ],
  "features": {
    "testing": {
      "description": "Dependencies needed for the tests.",
      "dependencies": [
        "catch2"
      ]
    }
  }
}
```

And then inside your CMakeLists.txt, if testing should be enabled tell vcpkg to enable the feature like follows:

```cmake
list(APPEND VCPKG_MANIFEST_FEATURES "testing")
```
